File: src/Structural_Patterns/Flyweight/flyweight.rb
# One real world example of the Flyweight design pattern in Ruby could be a text editor application. In this application, there may be a need to display a large amount of text on the screen at once, which could cause performance issues if each individual character was treated as a separate object.

# To solve this problem, the Flyweight pattern could be used to create a shared pool of character objects that can be reused throughout the application. Each character object would contain information about its position, font, and other properties, but the actual text content would be stored separately.

# Here is an example implementation of the Flyweight pattern in Ruby:

# ruby
class Character
  attr_accessor :position, :font

  def initialize(position, font)
    @position = position
    @font = font
  end

  def draw(content)
    # Draw the character on the screen using the shared font object
  end
end

class Font
  attr_accessor :name, :size

  def initialize(name, size)
    @name = name
    @size = size
  end
end

class CharacterFactory
  def initialize
    @characters = {}
  end

  def get_character(position, font_name, font_size)
    key = "#{position}-#{font_name}-#{font_size}"
    if @characters[key]
      # Return an existing character object from the pool
      @characters[key]
    else
      # Create a new character object and add it to the pool
      font = Font.new(font_name, font_size)
      character = Character.new(position, font)
      @characters[key] = character
      character
    end
  end
end

# Example usage:
factory = CharacterFactory.new
text = "Lorem ipsum dolor sit amet"
font_name = "Arial"
font_size = 12

text.each_char.with_index do |char, index|
  position = index * font_size
  character = factory.get_character(position, font_name, font_size)
  character.draw(char)
end


# In this example, the `Character` class represents a single character on the screen, and the `Font` class represents the font used to display the text. The `CharacterFactory` class is responsible for creating and managing a pool of shared `Character` objects based on their position and font properties.

# When the text is displayed on the screen, each character is retrieved from the factory using its position and font properties. If an existing character object with the same properties exists in the pool, it is returned. Otherwise, a new character object is created and added to the pool.

# By sharing objects in this way, the application can avoid creating unnecessary objects and improve performance when displaying large amounts of text.

File: src/Structural_Patterns/Flyweight/flyweight.ts
// One real world example of Flyweight design pattern code in Typescript can be used in a web application that displays a large number of images. Instead of creating a separate object for each image, the Flyweight pattern can be used to share common data between similar objects.

// For example, let's say we have an Image class that contains properties such as name, size, and source. Instead of creating a new object for each image, we can create a Flyweight factory that stores a pool of shared Image objects. When a new image is requested, the factory checks if an existing Image object with the same properties already exists in the pool. If it does, the factory returns the existing object. If not, it creates a new object and adds it to the pool for future use.

// Here's an example implementation of the Flyweight pattern in Typescript:


class Image {
  public name: string;
  public size: number;
  public source: string;

  constructor(name: string, size: number, source: string) {
    this.name = name;
    this.size = size;
    this.source = source;
  }
}

class ImageFlyweightFactory {
  private images: { [key: string]: Image } = {};

  public getImage(name: string, size: number, source: string): Image {
    const key = `${name}_${size}_${source}`;

    if (!this.images[key]) {
      this.images[key] = new Image(name, size, source);
    }

    return this.images[key];
  }
}

// Example usage:
const factory = new ImageFlyweightFactory();

const image1 = factory.getImage('cat', 100, 'https://example.com/cat.jpg');
const image2 = factory.getImage('dog', 200, 'https://example.com/dog.jpg');
const image3 = factory.getImage('cat', 100, 'https://example.com/cat.jpg');

console.log(image1 === image2); // false
console.log(image1 === image3); // true


// In this example, the ImageFlyweightFactory stores a pool of Image objects based on their name, size, and source properties. When a new Image object is requested, the factory checks if an existing object with the same properties already exists in the pool. If it does, the factory returns the existing object. If not, it creates a new object and adds it to the pool for future use.

// This implementation can be useful in a web application that displays a large number of images, as it reduces memory usage by sharing common data between similar objects.
